File: programmers/level 2/다리를_지나는_트럭.py
# 240418
def solution(bridge_length, weight, truck_weights):
    from collections import deque
    n = len(truck_weights)
    time = idx = 1
    on_bridge = truck_weights[0]
    end_time = deque()
    end_time.append((bridge_length, on_bridge))

    while end_time:

        if end_time[0][0] == time:
            on_bridge -= end_time.popleft()[1]

        if idx < n and on_bridge + truck_weights[idx] <= weight:
            end_time.append((time + bridge_length, truck_weights[idx]))
            on_bridge += truck_weights[idx]
            idx += 1

        time += 1

    return time

File: programmers/level 2/다리를_지나는_트럭.js
// 250718
function solution(bridge_length, weight, truck_weights) {
    const N = truck_weights.length
    let [time, bridge_weight, truck_i, bridge_i] = [2, truck_weights[0], 1, 0]
    const on_bridge = [[bridge_length+1, bridge_weight]]

    while (truck_i < N) {

        if (on_bridge[bridge_i][0] === time) {
            bridge_weight -= on_bridge[bridge_i][1]
            bridge_i += 1
        }

        const truck_weight = truck_weights[truck_i]

        if (truck_weight + bridge_weight <= weight) {
            on_bridge.push([time + bridge_length, truck_weight])
            bridge_weight += truck_weight
            truck_i += 1
        }

        time += 1
    }

    return on_bridge.at(-1)[0]
}
